File: packages/ogdf.js/src/layouts/misclayout/circular.js
import { PARAMETER_TYPE } from '../../utils/parameters'

//double minDistCC, double minDistCircle, double minDistLevel, double minDistSibling, double pageRatio
export default {
    minDistCC: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 20.0
    },
    minDistCircle: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 20.0
    },
    minDistLevel: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 20.0
    },
    minDistSibling: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 10.0
    },
    pageRatio: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1.0
    }
}

File: packages/ogdf.js/src/entry/misclayout/circular.cpp
#include <ogdf/misclayout/CircularLayout.h>
#include "../main.h"

EM_PORT_API(float *)
CIRCULAR(int node_num, int link_num, int *source, int *target, double minDistCC, double minDistCircle, double minDistLevel, double minDistSibling, double pageRatio)
{
    node *nodes;
    Graph G;
    GraphAttributes GA(G, GraphAttributes::nodeGraphics | GraphAttributes::edgeGraphics);

    nodes = new node[node_num];
    for (int i = 0; i < node_num; i++)
    {
        nodes[i] = G.newNode();
    }

    edge e;

    for (int i = 0; i < link_num; i++)
    {
        e = G.newEdge(nodes[source[i]], nodes[target[i]]);
        GA.bends(e);
    }
    //LayoutModule
    CircularLayout *model = new CircularLayout();

    model->minDistCC(minDistCC);
    model->minDistCircle(minDistCircle);
    model->minDistLevel(minDistLevel);
    model->minDistSibling(minDistSibling);
    model->pageRatio(pageRatio);

    model->call(GA);

    float *re = (float *)malloc(node_num * 2 * 4);
    for (int i = 0; i < node_num; ++i)
    {
        re[i * 2] = GA.x(nodes[i]);
        re[i * 2 + 1] = GA.y(nodes[i]);
    }

    return re;
}